cpp/tests/copying/gather_struct_tests.cu

```diff
@@ -0,0 +1,458 @@
+                                                  cudf::test::FloatingPointTypes,
+                                                  cudf::test::DurationTypes,
+                                                  cudf::test::TimestampTypes>;
+
+TYPED_TEST_CASE(TypedStructGatherTest, FixedWidthTypesNotBool);
+
+namespace {
+template <typename ElementTo, typename SourceElementT = ElementTo>
+struct column_wrapper_constructor {
+  template <typename ValueIter, typename ValidityIter>
+  auto operator()(ValueIter begin, ValueIter end, ValidityIter validity_begin) const
+  {
+    return cudf::test::fixed_width_column_wrapper<ElementTo, SourceElementT>{
+      begin, end, validity_begin};
+  }
+};
+
+template <>
+struct column_wrapper_constructor<std::string, std::string> {
+  template <typename ValueIter, typename ValidityIter>
+  cudf::test::strings_column_wrapper operator()(ValueIter begin,
+                                                ValueIter end,
+                                                ValidityIter validity_begin) const
+  {
+    return cudf::test::strings_column_wrapper{begin, end, validity_begin};
+  }
+};
+
+template <typename ElementTo, typename SourceElementT = ElementTo>
+auto get_expected_column(std::vector<SourceElementT> const& input_values,
+                         std::vector<bool> const& input_validity,
+                         std::vector<bool> const& struct_validity,
+                         std::vector<int32_t> const& gather_map)
+{
+  auto is_valid =  // Validity predicate.
+    [&input_values, &input_validity, &struct_validity, &gather_map](auto gather_index) {
+      assert(gather_index >= 0 && gather_index < gather_map.size() || "Gather-index out of range.");
+
+      auto i{gather_map[gather_index]};  // Index into input_values.
+
+      return (i >= 0 && i < static_cast<int>(input_values.size())) &&
+             (struct_validity.empty() || struct_validity[i]) &&
+             (input_validity.empty() || input_validity[i]);
+    };
+
+  auto expected_row_count{gather_map.size()};
+  auto gather_iter =
+    cudf::test::make_counting_transform_iterator(0, [is_valid, &input_values, &gather_map](auto i) {
+      return is_valid(i) ? input_values[gather_map[i]] : SourceElementT{};
+    });
+
+  return column_wrapper_constructor<ElementTo, SourceElementT>()(
+           gather_iter,
+           gather_iter + expected_row_count,
+           cudf::test::make_counting_transform_iterator(0, is_valid))
+    .release();
+}
+}  // namespace
+
+TYPED_TEST(TypedStructGatherTest, TestSimpleStructGather)
+{
+  using namespace cudf::test;
+
+  // Testing gather() on struct<string, numeric, bool>.
+
+  // 1. String "names" column.
+  auto names = std::vector<std::string>{"Vimes", "Carrot", "Angua", "Cheery", "Detritus", "Slant"};
+  auto names_validity = std::vector<bool>{1, 1, 1, 1, 1, 1};
+  auto names_column   = strings_column_wrapper{names.begin(), names.end(), names_validity.begin()};
```

Comment:
Many opportunities for `const` here :)

Reply:
I have switched to using `const` where viable.